### PR TITLE
Fixes printing issue with TensorWrapper

### LINKED ISSUE #382: Printing tensor broken in vmap(grad( case
Repro:
```
import torch
import torch.nn as nn
from functorch import vmap, grad, make_functional

class TestNetwork(nn.Module):
  def __init__(self):
    super().__init__()
    self.net = nn.Linear(5, 5)
  def forward(self, x):
    x = self.net(x)
    print(x)
    print(x.shape)
    print("Should be (10,5)")
    x[:, 0] += 3
    return x

model = TestNetwork()
data = torch.randn(10,5)
targets = torch.randn(10, 5)

func_model, params = make_functional(model)

def compute_loss(params, data, targets):
    preds = func_model(params, data.unsqueeze(0))
    return torch.mean((preds - targets.unsqueeze(0)) ** 2)

# This seems fine.
grad(compute_loss)(params, data, targets)

# This breaks
per_sample_grads = vmap(grad(compute_loss), (None, 0, 0))(params, data, targets)
```
gives `RuntimeError: Cannot access data pointer of Tensor that doesn't have storage`

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Hacky fix to fix printing tensors inside `grad`
- other nits and cosmetics

Fixes #382

@zou3519 what do you think ?

### CODE REVIEW

File: functorch/__init__.py [hunk 1]
```diff
@@ -91,7 +91,15 @@ def _functorch_str(tensor):
...
     value = _C.get_unwrapped(tensor)
+    # Disable temporarily kDynamicLayerFrontModeKey/kDynamicLayerBackModeKey as included dispatch keys
+    dl_enabled = _C.tls_set_is_included()
+    if (dl_enabled):
+        _C._set_dynamic_layer_keys_included(False)
     value_repr = repr(value)
+    # Reenable kDynamicLayerFrontModeKey/kDynamicLayerBackModeKey as included dispatch keys
+    if (dl_enabled):
+        _C._set_dynamic_layer_keys_included(True)
+
```
Comment (zou3519) [inline functorch/__init__.py:102]: Ugh I think I introduced this bug last month. I agree on the approach (that when we're printing the repr we should disable all things functorch related). Let me think about this a little more and get back to you

Comment (zou3519) [inline functorch/__init__.py:102]: @vfdev-5 I'm a bit confused about why our current tests pass ([1](https://github.com/pytorch/functorch/blob/b2d176daec0cfaee7d047a77d63cfc1ef020e057/test/test_vmap.py#L986-L1002), [2](https://github.com/pytorch/functorch/blob/main/test/test_eager_transforms.py#L723-L740)). Do you have some insight into that?

File: functorch/__init__.py [hunk 2]
```diff
@@ -91,7 +91,15 @@ def _functorch_str(tensor):
...
 
     value = _C.get_unwrapped(tensor)
+    # Disable temporarily kDynamicLayerFrontModeKey/kDynamicLayerBackModeKey as included dispatch keys
+    dl_enabled = _C.tls_set_is_included()
+    if (dl_enabled):
+        _C._set_dynamic_layer_keys_included(False)
     value_repr = repr(value)
+    # Reenable kDynamicLayerFrontModeKey/kDynamicLayerBackModeKey as included dispatch keys
+    if (dl_enabled):
+        _C._set_dynamic_layer_keys_included(True)
```
Comment (zou3519) [inline functorch/__init__.py:101]: We probably want to use some "RAII" here: if the `repr` call fails then we want to make sure the state is still OK:
```
try: 
  if (dl_enabled):
        _C._set_dynamic_layer_keys_included(False)
finally:
   if (dl_enabled):
        _C._set_dynamic_layer_keys_included(True)
```

File: functorch/__init__.py (hunk 1 above)
Comment (zou3519) [reply functorch/__init__.py:102]: I see, the problem is the printing path does different things on large tensor vs small tensors

**Review by zou3519 — COMMENTED** [review]
Approach LGTM. One comment about RAII. We should also beef up our testing of printing and test some more cases :) the previous tests I wrote are not comprehensive at all.

File: functorch/__init__.py (hunk 1 above)
Comment (zou3519) [reply functorch/__init__.py:102]: I thought about it and it looks like this was a pre-existing condition (and not caused by my refactor of DynamicLayer last month)

Comment (vfdev-5) [reply functorch/__init__.py:102]: Yes, I'm also surprised that test_eager_transforms.py::test_tensor_print was passing before this PR. I'll check that.
As for vmap, IMO they should always pass as it does not modify dispatch keys (tls_set_dispatch_key_included).

File: functorch/__init__.py (hunk 2 above)
Comment (vfdev-5) [reply functorch/__init__.py:101]: I see your point, but just wonder if we really need to do that. If `value_repr = repr(value)` fails then script execution stops. If user does try-except some part of his/her code where `print(t)` errors, catches the error and continue the script, we should ensure that the state is OK and your pattern would totally make sense. I wonder how "typical" this pattern in python development ?

Comment (zou3519) [reply functorch/__init__.py:101]: > if user does try-except some part of his/her code where print(t) errors, catches the error and continue the script, we should ensure that the state is OK and your pattern would totally make sense.  I wonder how "typical" this pattern in python development ?

You're right that this pattern would probably not be common in user code. There's two concrete use cases for this that I've seen:
1. Someone has a long-running training loop. If an error occurs, I've seen some people do a try-catch and then save their model state in the finally() step so that no training progress is lost.
2. Testing. If someone (really, just us :P) has a test suite that tests tensor printing, if the printing raises an error in a unit test then it'll affect the state of all other subsequent unit tests

File: functorch/__init__.py (hunk 1 above)
Comment (vfdev-5) [reply functorch/__init__.py:102]: Seems like there is no printing issue for scalars but there is an issue with non-scalar tensors.

File: test/test_eager_transforms.py [hunk 3]
```diff
@@ -720,22 +720,22 @@ def foo(x):
         x = torch.tensor(3.14, device=device)
         functorch.grad(foo)(x)
 
-    @onlyCPU
     def test_tensor_print(self, device):
-        x = torch.tensor(3.14, device=device)
+        x = torch.tensor([[3.14, ], [3.14, ]], device=device)
```
Comment (zou3519) [inline test/test_eager_transforms.py:724]: Could we keep the original tests and add some new test cases? It's probably valuable to test a variety of tensors to make sure they can be printed:
- torch.tensor(3.14, device=device)
- torch.tensor([[3.14, ], [3.14, ]], device=device)
- Some very big tensor (`torch.randn(64, 3, 32, 32)`). We can probably compare that `repr(t_inside_transform)` is `GradTrackingTensor(lvl=2, value={repr(t)}` or something

**Review by zou3519 — COMMENTED** [review]
The code LGTM, but we should try to add some more rigorous tests. I suggested adding some cases that include the cross product of { different tensors } x { vmap, grad, vmapxvmap, gradxgrad, vmapxgrad } transforms

File: test/test_eager_transforms.py (hunk 3 above)
Comment (zou3519) [reply test/test_eager_transforms.py:724]: There are probably other edge cases that we want to test that I'm not thinking of here

File: test/test_eager_transforms.py [hunk 4]
```diff
@@ -720,24 +719,43 @@ def foo(x):
...
-                )""")
-        self.assertEqual(buf, expected)
+    @parametrize("op_list_data", [
+        subtest(([vmap, ], [(64, 3, 32, 32), (4, 2)]), name='vmap'),
+        subtest(([vmap, vmap], [(64, 3, 32, 32), (4, 3, 2)]), name='vmap_vmap'),
+        subtest(([grad, ], [(0, ), [], (4, 2), (64, 3, 32, 32)]), name='grad'),
+        subtest(([grad, grad], [[], ]), name='grad_grad'),
+        subtest(([vmap, grad], [(4, 2)]), name='vmap_grad'),
+    ])
+    def test_tensor_print(self, device, op_list_data):
```
Comment (vfdev-5) [inline test/test_eager_transforms.py:729]: @zou3519 how about that ? We can not make data x op product as data depends on the op.

File: test/test_vmap.py [hunk 5]
```diff
@@ -984,24 +983,6 @@ def f(x, y):
...
-            return x
-
-        vmap(vmap(foo))(x)
-        expected = textwrap.dedent("""\
-                BatchedTensor(lvl=3, bdim=0, value=
-                    BatchedTensor(lvl=2, bdim=0, value=
-                        tensor([[3.1400]])
-                    )
-                )""")
-        self.assertEqual(buf, expected)
```
Comment (zou3519) [inline test/test_vmap.py:1003]: Let's add a test for different bdims here (to make sure we didn't hardcode bdim=0 lol)

File: test/test_eager_transforms.py [hunk 6]
```diff
@@ -720,24 +719,43 @@ def foo(x):
...
+
+        op_list, shapes = op_list_data
+        data = [torch.randn(s, device=device) for s in shapes]
```
Comment (zou3519) [inline test/test_eager_transforms.py:732]: Can we also test different dtypes somehow?

**Review by zou3519 — APPROVED** [review]
Some minor comments about testing different dtypes / bdims. This looks pretty good though

File: test/test_eager_transforms.py (hunk 4 above)
Comment (zou3519) [reply test/test_eager_transforms.py:729]: This looks great, thanks!

**Comment by vfdev-5** [comment]
@zou3519 I added bdim and dtypes in the tests. Anything else do you think we can need of ?

**Review by zou3519 — APPROVED** [review]
Nah this looks great, thank you!
